test/distributed/org/apache/cassandra/distributed/test/CleanupFailureTest.java

```diff
@@ -0,0 +1,153 @@
+import org.apache.cassandra.distributed.api.IInvokableInstance;
+import org.apache.cassandra.distributed.test.ring.BootstrapTest;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.service.StorageService;
+import org.mortbay.util.IO;
+
+import static java.util.Arrays.asList;
+import static org.apache.cassandra.distributed.action.GossipHelper.bootstrap;
+import static org.apache.cassandra.distributed.action.GossipHelper.decommission;
+import static org.apache.cassandra.distributed.action.GossipHelper.pullSchemaFrom;
+import static org.apache.cassandra.distributed.action.GossipHelper.statusToBootstrap;
+import static org.apache.cassandra.distributed.action.GossipHelper.withProperty;
+import static org.apache.cassandra.distributed.api.Feature.GOSSIP;
+import static org.apache.cassandra.distributed.api.Feature.NATIVE_PROTOCOL;
+import static org.apache.cassandra.distributed.api.Feature.NETWORK;
+
+public class CleanupFailureTest extends TestBaseImpl
+{
+    private static Cluster CLUSTER;
+
+    @BeforeClass
+    public static void before() throws IOException
+    {
+        CLUSTER = init(Cluster.build()
+                              .withNodes(3)
+                              .withConfig(config -> config.with(NETWORK, GOSSIP, NATIVE_PROTOCOL))
+                              .start());
+    }
+
+    @AfterClass
+    public static void after()
+    {
+        if (CLUSTER != null)
+            CLUSTER.close();
+    }
+
+    @Test
+    public void testCleanupFailsDuringOngoingDecommission()
+    {
+        // set up keyspace and table
+        CLUSTER.schemaChange("CREATE KEYSPACE IF NOT EXISTS " + KEYSPACE + " WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'};");
+        CLUSTER.schemaChange("CREATE TABLE IF NOT EXISTS " + KEYSPACE + ".tbl (pk int, ck int, v int, PRIMARY KEY (pk, ck))");
+        CLUSTER.schemaChange("ALTER KEYSPACE " + KEYSPACE + " WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'};");
+        CLUSTER.schemaChange("ALTER KEYSPACE system_distributed WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'};");
+
+        // disable autocompaction
+        CLUSTER.get(1).nodetoolResult("disableautocompaction", KEYSPACE).asserts().success();
+
+        // populate data and flush
+        for(int i=0; i < 20; i++){
+            CLUSTER.get(1).coordinator().execute("INSERT INTO " + KEYSPACE + ".tbl (pk, ck, v) VALUES (?, ?, ?)",
+                    ConsistencyLevel.ALL,
+                    1, i, i);
+            CLUSTER.get(1).flush(KEYSPACE);
+        }
+
+        // assert data has been populated
+        Object[][] beforeDecommResponse = CLUSTER.get(1).executeInternal("SELECT * FROM " + KEYSPACE + ".tbl ;");
+        Assert.assertEquals(20, beforeDecommResponse.length);
+
+        // assert 20 sstables
```

Comment:
I don't think we need 20 sstables to reproduce this test, as it seems to be reproducible with a single sstable. To make this simpler I think we don't need to flush after inserting each row nor assert the number of sstables.